lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -0,0 +1,850 @@
+
+        Args:
+            csr (bytes): Certificate Signing Request
+
+        Returns:
+            None
+        """
+        logger.info("Received request to create certificate")
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            message = (
+                f"Relation {self.relationship_name} does not exist - "
+                f"The certificate request can't be completed"
+            )
+            logger.error(message)
+            raise RuntimeError(message)
+        relation_data = load_unit_relation_data(relation.data[self.model.unit])
+        new_certificate_request = {"certificate_signing_request": csr.decode()}
+        if "certificate_signing_requests" in relation_data:
+            certificate_request_list = relation_data["certificate_signing_requests"]
+            if new_certificate_request in certificate_request_list:
+                logger.info("Request was already made - Doing nothing")
+                return
+            certificate_request_list.append(new_certificate_request)
+        else:
+            certificate_request_list = [new_certificate_request]
+        relation.data[self.model.unit]["certificate_signing_requests"] = json.dumps(
+            certificate_request_list
+        )
+        logger.info("Certificate request sent to provider")
+
+    def renew_certificate(self, certificate_signing_request: bytes) -> None:
+        """Renews Certificate.
+
+        Revokes certificate than requests one with the same CSR.
+
+        Args:
+            certificate_signing_request:
+
+        Returns:
+            None
+        """
+        self.revoke_certificate(certificate_signing_request)
+        self.request_certificate(certificate_signing_request)
+
+    def revoke_certificate(self, certificate_signing_request: bytes) -> None:
+        """Removes CSR from relation data.
+
+        The provider of this relation is then expected to remove certificates associated to this
+        CSR from the relation data as well and emit a revoke_certificate event for the provider
+        charm to interpret.
+
+        Args:
+            certificate_signing_request (bytes): Certificate Signing Request
+
+        Returns:
+            None
+        """
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            raise RuntimeError(f"Relation {self.relationship_name} does not exist")
+        relation_data = load_unit_relation_data(relation.data[self.model.unit])
+        certificate_request_list = relation_data.get("certificate_signing_requests")
+        if not certificate_request_list:
+            logger.info("No CSR in relation data.")
+            return
+        for i in range(len(certificate_request_list)):
+            if (
+                certificate_request_list[i]["certificate_signing_request"]
+                == certificate_signing_request.decode()
+            ):
+                certificate_request_list.pop()
```

Comment:
I think this would be more pythonic, however, if you do not take my suggestion, there seems to be a bug on line 769 as you are not passing the index to `pop()`.

```suggestion
        for csr in certificate_request_list[:]:
            if (
                csr["certificate_signing_request"]
                == certificate_signing_request.decode()
            ):
                certificate_request_list.remove(csr)
```

Reply:
Agreed, fixed